samples/bluetooth/iso_combined_bis_and_cis/src/combined_bis_cis.c

```diff
@@ -0,0 +1,308 @@
+ */
+
+#include <zephyr/kernel.h>
+#include <zephyr/bluetooth/bluetooth.h>
+#include <zephyr/bluetooth/conn.h>
+#include <zephyr/bluetooth/iso.h>
+#include <zephyr/sys/ring_buffer.h>
+
+#include "iso_combined_bis_and_cis.h"
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(app_bis_cis, LOG_LEVEL_INF);
+
+#define ADV_NAME_STR_LEN (sizeof(CONFIG_BT_DEVICE_NAME))
+
+#define SDU_INTERVAL_US			(10000u) /* 10 ms */
+#define ACL_INTERVAL_BLE_UNITS		(16u)	 /* 20 ms in 1.25 ms BLE units*/
+#define PERIODIC_ADV_INTERVAL_BLE_UNITS (16u)	 /* 20 ms in 1.25 ms BLE units*/
+#define RETRANSMISSION_NUMBER		(10u)
+#define MAX_TRANSPORT_LATENCY_MS	(20u)
+#define SDU_RING_BUFFER_SIZE		(4u)
+
+/* Ring buffer contains length of SDU before each SDU, so allocate 1 additional byte */
+#define RING_BUF_ELEMENT_SIZE (1 + CONFIG_BT_ISO_TX_MTU)
+RING_BUF_DECLARE(sdu_rb, SDU_RING_BUFFER_SIZE * (RING_BUF_ELEMENT_SIZE));
+static uint8_t rcvd_sdu_cnt;
+
+static void scan_start(void)
+{
+	int err;
+
+	err = bt_le_scan_start(
+		BT_LE_SCAN_PARAM(BT_LE_SCAN_TYPE_ACTIVE, BT_LE_SCAN_OPT_FILTER_DUPLICATE,
+				 BT_GAP_SCAN_FAST_INTERVAL, BT_GAP_SCAN_FAST_INTERVAL),
+		NULL);
+	if (err) {
+		LOG_ERR("Scanning failed to start (err %d)", err);
+		return;
+	}
+}
+
+static struct bt_iso_chan *free_iso_chan_get(void)
+{
+	struct bt_iso_chan **iso_channels;
+
+	iso_channels = iso_rx_channels_get();
+
+	struct bt_iso_chan *iso_chan = NULL;
+
+	if (iso_channels[0]->state == BT_ISO_STATE_DISCONNECTED) {
+		iso_chan = iso_channels[0];
+	}
+
+	return iso_chan;
+}
+
+static bool adv_data_parse_cb(struct bt_data *data, void *user_data)
+{
+	char *name = user_data;
+	uint8_t len;
+
+	switch (data->type) {
+	case BT_DATA_NAME_SHORTENED:
+	case BT_DATA_NAME_COMPLETE:
+		len = MIN(data->data_len, ADV_NAME_STR_LEN - 1);
+		memcpy(name, data->data, len);
+		name[len] = '\0';
+		return false;
+	default:
+		return true;
+	}
+}
+
+static void scan_recv(const struct bt_le_scan_recv_info *info, struct net_buf_simple *buf)
+{
+	char name_str[ADV_NAME_STR_LEN] = {0};
+
+	bt_data_parse(buf, adv_data_parse_cb, name_str);
+
+	if (strncmp(name_str, CONFIG_BT_DEVICE_NAME, ADV_NAME_STR_LEN) != 0) {
+		return;
+	}
+
+	if (bt_le_scan_stop()) {
+		return;
+	}
+
+	const struct bt_le_conn_param *conn_param =
+		BT_LE_CONN_PARAM(ACL_INTERVAL_BLE_UNITS, ACL_INTERVAL_BLE_UNITS, 0, 400);
+
+	struct bt_conn *conn;
+	int err = bt_conn_le_create(info->addr, BT_CONN_LE_CREATE_CONN, conn_param, &conn);
+
+	if (err) {
+		LOG_ERR("Create conn to %s failed (%d)", name_str, err);
+	}
+}
+
+static struct bt_le_scan_cb scan_callbacks = {
+	.recv = scan_recv,
+};
+
+static void process_rcvd_sdu_cb(struct net_buf *buf)
+{
+	uint8_t sdu_buffer[CONFIG_BT_ISO_TX_MTU];
+	uint8_t sdu_len = (uint8_t)buf->len;
+
+	memcpy(sdu_buffer, buf->data, sdu_len);
+
+	/* write length of SDU to ring-buffer */
+	uint32_t wrote = ring_buf_put(&sdu_rb, &sdu_len, 1);
+
+	__ASSERT_NO_MSG(wrote == 1);
```

Comment:
Not entirely sure how the ring buffer is implemented, but is there a chance that the data of an SDU overwrites the length of a previous SDU? Or is that not allowed?
I.e. will the ring buffer overwrite existing non-pulled data or will it fail to add more data if it's full?